google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -104,6 +112,63 @@ def _wrap_callback_errors(
         on_callback_error(exc)
 
 
+def _get_ack_errors(exc: exceptions.GoogleAPICallError) -> Optional["containers.ScalarMap"]:
+    if not exc.response:
+        _LOGGER.debug("No response obj in errored RPC call.")
+        return None
+    try:
+        status = rpc_status.from_call(exc.response)
+        if not status:
+            _LOGGER.debug("Unable to get status of errored RPC.")
+            return None
+        for detail in status.details:
+            if detail.Is(ErrorInfo.DESCRIPTOR):
+                info = ErrorInfo()
+                if not detail.Unpack(info):
+                    _LOGGER.debug("Unable to unpack ErrorInfo.")
+                    return None
+                return info.metadata
+    # Possible "If the gRPC call’s code or details are inconsistent
+    # with the status code and message inside of the
+    # google.rpc.status.Status"
+    except ValueError:
+        _LOGGER.debug( "ValueError when parsing ErrorInfo.", exc_info=True)
+        return None
+
+def _process_futures(future_reqs_dict: "containers.ScalarMap",
+                    errors_dict: Optional["containers.ScalarMap"]):
+    """Process futures by referring to errors_dict.
+
+    The errors returned by the server in `errors_dict` are used to complete
+    the request futures in `future_reqs_dict` (with a success or exception) or
+    to return requests for further retries.
+    """
+    requests_completed = []
+    requests_to_retry = []
+    for ack_id in future_reqs_dict:
+        if errors_dict and ack_id in errors_dict:
+            exactly_once_error = errors_dict[ack_id]
+            if exactly_once_error.startswith("PERMANENT_"):
+                exc = RuntimeError("Permanent error: " + exactly_once_error)
```

Comment:
I am not sure I understand how this exception is handled. Can you explain? Making sure this will not crash the program.

Reply:
The future is completed with an exception. If the user calls the future's result() method, it'll throw an exception. They should check if the future is done successfully first. This is standard behavior for futures in Python.